Documentation on how to make a boefje, normalizer, model, bit and report with examples.

Souf149 opened the pull request:
Wrote documentation on how to make your own boefjes, normalizers, models, bits and reports. This sounds useful for people wanting to start developing on their own. 

### Changes

Added:
- New plugins
- New OOI-model
- New report type
- New documentation page

The new plugins, OOI-model and report type are made as examples for the documentation.

### Issue link

Closes https://github.com/minvws/nl-kat-coordination/issues/2966

### Demo

![image](https://github.com/minvws/nl-kat-coordination/assets/34127049/d0303f0a-1267-406e-999b-803523367d31)
![image](https://github.com/minvws/nl-kat-coordination/assets/34127049/c43aece7-827f-40ad-a21a-61f9ebe3b50c)

---

Note:
The commits are GPG-signed, does that count? 😅


### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [x] I have written unit tests for the changes or fixes I made.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication

- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have made corresponding changes to the documentation, if necessary.
- [x] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---

## Checklist for code reviewers:

- [ ] The code does not violate Model-View-Template and our other architectural principles.
- [ ] The code prioritizes readability over performance where appropriate.
- [ ] The code does not bypass authentication or security mechanisms.
- [ ] The code does not introduce any dependency on a library that has not been properly vetted.
- [ ] The code contains docstrings, comments, and documentation where needed.

---

## Checklist for QA:

- [ ] I have checked out this branch, and successfully ran a fresh `make reset`.
- [ ] I confirmed that there are no unintended functional regressions in this branch:
- [ ] I have managed to pass the onboarding flow
- [ ] Objects and Findings are created properly
- [ ] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:

- _bullet point + screenshot (if useful) per tested functionality_

### What doesn't work:

- _bullet point + screenshot (if useful) per tested functionality_

### Bug or feature?:

- _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._

Souf149 commented:
Made the hello katty boefje use the `base.Dockerfile` instead of its own Dockerfile. Requested by @Donnype

ammar92 reviewed (commented):
This looks great, many thanks for that 👍 It covers most of the plugin and report development and serves as a good guide for this purpose.

However, I believe the implementation of the plugin isn't necessary. Introducing a fictive plugin could raise several questions and potential migration issues, particularly with the models. Good documentation, like you currently have with code examples and file tree structures, should be sufficient.

I suggest removing the code changes (plugin files, test and makefiles). After that there's only a bit of polishing work needed for the documentation to make this even better, such as:
- Fix some spelling mistakes, ReST syntax and ensuring write style is consistent with the rest of the documentation
- You can reference `is_ipv4` being used, but diving into this is too specific imho
- In some of the points you mention a *raw meta file* and *return file*, this should be: *meta file* and *raw file*
- I have a few more suggestions that I will provide in the next iteration

Souf149 commented:
@ammar92, Could you elaborate with what you meant with `Fix some spelling mistakes, ReST syntax`? I assume that you meant Markdown syntax, so I have added language specification to the code-blocks to better comply with the Markdown ruleset. And I have changed the structure at the start of the documentation to better follow the style of the existing documentation. 

I hope these changes are more to your likings, I'd love to hear what more you have as feedback 😄

ammar92 reviewed (dismissed):
Looks good, no more remarks

stephanie0x00 commented on docs/source/developer_documentation/development_tutorial/index.rst:
```diff
@@ -0,0 +1,32 @@
++------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| Normalizer | A plugin that listens to specified boefjes' raw data, and creates new OOIs from the data that they find. This is often called a whisker.                                                                                                                              |
++------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| Bit        | A plugin that waits for specified OOIs and creates more OOIs from these (mostly used to create findings.)                                                                                                                                                             |
```
```suggestion
| Bit        | A plugin that waits for specified OOIs and creates more OOIs from these (mostly used to create findings).                                                                                                                                                             |
```

stephanie0x00 commented on docs/source/developer_documentation/development_tutorial/index.rst:
```diff
@@ -0,0 +1,32 @@
++------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| Bit        | A plugin that waits for specified OOIs and creates more OOIs from these (mostly used to create findings.)                                                                                                                                                             |
++------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| Finding    | A special OOI that tells the user certain events have happened that might interest them. For example, a Finding could say that the server's ssh port is open while it should not.                                                                                     |
```
```suggestion
| Finding    | A special OOI that tells the user certain events have happened that might interest them. For example, a Finding could say that the server's SSH port is open while it should not.                                                                                     |
```

stephanie0x00 commented on docs/source/developer_documentation/development_tutorial/creating_a_new_model.md:
````diff
@@ -0,0 +1,67 @@
+from octopoes.models.ooi.greeting import Greeting
+```
+
+And then adding our Greeting model to the `ConcreteOOIType` set.
````
Can you describe how/where to do this?

stephanie0x00 commented on docs/source/developer_documentation/development_tutorial/creating_a_boefje.md:
````diff
@@ -0,0 +1,146 @@
+	docker build -f images/base.Dockerfile -t openkat/hello-katty --build-arg BOEFJE_PATH=./boefjes/plugins/kat_hello_katty .
+```
+
+This was the creation of our first boefje. If we run OpenKAT now we should be able to see this boefje sitting in the KAT-alogus. Let’s try it out!
````
I think you miss the reference here to the next part of the tutorial on creating the model? It is not referenced in the other tutorial pages.

stephanie0x00 reviewed (commented):
Very nice tutorial!! Some small suggestions can be found below.

Souf149 commented:
Thank you! I will look at the suggested changes and clarify my choices. 😊

Souf149 commented on docs/source/developer_documentation/development_tutorial/creating_a_boefje.md:
````diff
@@ -0,0 +1,146 @@
+}
+```
+
+- **`id`**: A unique identifier for the boefje.
````
I have not encountered a problem with the names I have used for the boefjes' `id`. There might be some rules the OpenKAT team might want to decide on (such as all of it having to be lowercase), but these currently do not exist.

Souf149 replied:
I'd say for now this is sufficient information for a developer, but naming conventions are something that can be looked for in the future.

Souf149 commented on docs/source/developer_documentation/development_tutorial/creating_a_boefje.md:
````diff
@@ -0,0 +1,146 @@
+    ]
+```
+
+The most important part is the return value we send back:
````
Indeed something got went wrong here, it has been added back in.

Souf149 commented on docs/source/developer_documentation/development_tutorial/creating_a_boefje.md:
```diff
@@ -0,0 +1,146 @@
+
+The most important part is the return value we send back:
+The final task of creating a boefje is specifying what DockerFile our boefje should use. We can do this inside the file located in `boefjes/Makefile`.
+Inside the `images` rule. We have to specify that we want to use `base.Dockerfile` since our boefje does not require any special packages to run. This is as simple as adding a single line. Here is what that would look like in our case:
```
It is quite convoluted to make that change. I think the best solution would be to have a separate page explaining how to do that, or we could append it to the current page. Which one do you think is better here?

underdarknl replied:
Requirements/deps are going to be tracker per plugin in their respective directory / repository. They will end up being build into the container images and as such will be documented later on for this new design.

stephanie0x00 commented:
Looks good @Souf149! Very nice tutorial. :heart: Ready for merge.
